anvio/kegg.py

```diff
@@ -1518,12 +1776,32 @@ def store_annotations_in_db(self):
                 gene_function_calls_table.create(self.kegg_module_names_dict)
             if self.kegg_module_classes_dict:
                 gene_function_calls_table.create(self.kegg_module_classes_dict)
+            if self.kegg_brite_categorizations_dict:
+                gene_function_calls_table.create(self.kegg_brite_categorizations_dict)
         else:
             self.run.warning("There are no KOfam hits to add to the database. Returning empty handed, "
                              "but still adding KOfam as a functional source.")
             gene_function_calls_table.add_empty_sources_to_functional_sources({'KOfam'})
 
 
+    def store_brite_info_in_db(self):
+        """Fills in the table of BRITE hierarchy information in the contigs database.
+
+        The only data stored in this table regards hierarchy depths, gleaned from the full set of
+        information in the modules database. Depth data is needed for pruning categorizations to the
```

Comment:
If every contigs database would store the same hierarchy depth information, wouldn't it be better to just make this another table in the modules database instead? That way we would only increase one database version and different contigs DBs wouldn't be storing redundant information (and would be smaller). 

But I might be misunderstanding something here, so please tell me if there is a reason to have this info in the contigs db :)

Reply:
Very good point :) I think this is worth thinking about!

Reply:
It's funny when @ivagljiva reads my mind and grasps nuances of implementation that I mulled over myself. The reason I didn't store hierarchy depth information in the modules database is that operations on the contigs database shouldn't depend on the existence of the modules database. I envision the depth data coming into use when displaying categories in `anvi-display-functions`. If the user wants to display more general categories, lumping together the N categories at the bottom of the hierarchies, then depths need to be stored somewhere, and ideally this should not require a call to the modules database that originally categorized functions in the contigs database. The thought of storing depth data as an anvi'o constant also flitted through my mind, but hierarchy depths can surely change in the future.

I now see this might have implications for module completeness. Should module completeness calculations require referring from the contigs to the modules database? Or should module completeness be stored directly in the contigs database at the time of module annotation.

Reply:
> Should module completeness calculations require referring from the contigs to the modules database?

AFAIK, it already does. You need to have the modules-db properly setup on your computer to run completion stats based on KOfam hits in a given contigs-db. But not having it does not prevent us from distributing a contigs-db alone.

It would make a lot of sense to keep BRITE categories or hierarchy depth information only in modules-db, rather than the contigs-db. But this is going to add a lot of technical headaches (but will be much more elegant).

First of all, every time we run "get gene functions", we will have to add BRITE data on the fly to the resulting functions dict based on KOfam hits we find (i.e., get all functions will return more sources than what is literally in the gene functions table). I am not sure if this is a serious computational bottleneck, but I wouldn't think so.

Then there is the issue of reproducibility / interoperability. For instance, imagine this scenario:

A user downloads a contigs-db from a paper. It has KOfam annotations, but they don't have a modules-db setup on their computer. Well, in that. case they will not be able to get the BRITE hierarchy information for the genes in that contigs-db. There will be a warning or an error saying that they should run `anvi-setup-kegg-kofams` first. I'm not sure if this is a problem either because the same user will not be able to generate metabolism completion estimates without a modules-db on their system, either.

An alternative is to have this 'on-the-fly' / modules-db requiring behavior the default, and also have a flag for the program `anvi-setup-kegg-kofams`, `--store-brite-hierarchy-data-in-contigs-db` where these hierarchies would be literally stored in contigs-db. In this case the same user above would be able to work with BRITE hierarchies without having to setup a modules-db on their computer if the contigs-db was annotated with that flag. But I'm not sure if this level of complexity is justified since those who will use BRITE hierarchies will likely also have modules-db setup on their computers. It is just an issue of helpful warning / error messages so users can find their ways by themselves.

What do you guys think of all this?

Reply:
I'm in favor of keeping BRITE categories in the gene functions table of the contigs db since KOfam module assignments are stored there. To guarantee that the BRITE and module categorizations are from the same version of KEGG, they should either both be in the contigs db or both be in the modules db, since the modules db can be changed.

@meren's proposed flag to store BRITE categorizations in the contigs db provides a solution to this problem by "locking" the version in place. Anvi'o could also emit a warning if the modules db hash is different than the hash recorded in the contigs db.

On a related note, it is probably worth recording a timestamp for KEGG in the modules db when the db is set up with a downloaded or archived version of KEGG so that the user can reference the version used in the analysis.

The issue of database dependencies could be addressed by a tabular log file listing the paths and hashes of dbs associated with the contigs db that anvi'o generates and modifies as different commands are run. The file would appear in the same directory as the contigs db upon creation. If `anvi-run-kegg-kofams` is then run, a line would be added showing that the contigs db is associated with the modules db. If `anvi-profile` is then run, another line would be added with this historical record.

Reply:
After talking to @meren, I better understand the rationale for confining BRITE hierarchies to the modules db and categorizing KOfams from the contigs db on the fly. We warn the user to keep KEGG annotations in the contigs db up-to-date with the modules db. The user shouldn't have a modules db from a version of KEGG ahead of the contigs db annotations.

So how should BRITE categorizations be found on the fly? `anvi-interactive` pulls functions from the contigs db using `dbops.ContigsSuperclass.init_functions`, populating a dictionary of function calls from different sources. If KOfams are present, the modules db can be loaded and BRITE categorizations added to this dictionary. There should be a method of `kegg.ModulesDatabase` that takes a list of KOfams and returns BRITE categorizations in a dictionary compatible with the gene functions format.

Reply:
I decided to completely get rid of the separate table for BRITE hierarchy depths, since we decided against including this as "static" information in the contigs db, and depths can be computed on the fly from the modules db for `anvi-display-functions` or another application.